gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KotlinCompilerRunner.kt

```diff
@@ -0,0 +1,113 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@file:Suppress("INVISIBLE_REFERENCE", "INVISIBLE_MEMBER")
+
+package com.google.devtools.ksp.gradle
+
+import org.gradle.api.Task
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.ListProperty
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Classpath
+import org.gradle.api.tasks.Internal
+import org.jetbrains.kotlin.cli.common.arguments.K2JSCompilerArguments
+import org.jetbrains.kotlin.cli.common.arguments.K2JVMCompilerArguments
+import org.jetbrains.kotlin.cli.common.arguments.K2MetadataCompilerArguments
+import org.jetbrains.kotlin.cli.common.arguments.K2NativeCompilerArguments
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy
+import org.jetbrains.kotlin.gradle.utils.newInstance
+import java.io.File
+
+interface KotlinCompilerRunner {
+    @get:Internal
+    val compilerExecutionStrategy: Property<KotlinCompilerExecutionStrategy>
+
+    @get:Internal
+    val useDaemonFallbackStrategy: Property<Boolean>
+
+    @get:Internal
+    val kotlinDaemonJvmArguments: ListProperty<String>
+
+    @get:Classpath
+    val compilerClasspath: ConfigurableFileCollection
+}
+
+interface KotlinJvmCompilerRunner : KotlinCompilerRunner {
+    fun runJvmCompilerAsync(
+        args: K2JVMCompilerArguments,
+        sources: List<File>,
+        commonSources: List<File>,
+        outputs: List<File>
+    )
+}
+
+interface KotlinJsCompilerRunner : KotlinCompilerRunner {
+    fun runJsCompilerAsync(
+        args: K2JSCompilerArguments,
+        sources: List<File>,
+        commonSources: List<File>,
+        outputs: List<File>
+    )
+}
+
+interface KotlinMetadataCompilerRunner : KotlinCompilerRunner {
+    fun runMetadataCompilerAsync(
+        args: K2MetadataCompilerArguments,
+        sources: List<File>,
+        commonSources: List<File>,
+        outputs: List<File>
+    )
+}
+
+interface KotlinNativeCompilerRunner : KotlinCompilerRunner {
+    fun runNativeCompilerAsync(
+        args: K2NativeCompilerArguments,
+        sources: List<File>,
+        commonSources: List<File>,
+        outputs: List<File>,
+    )
+}
+
+fun createKotlinJvmCompilerRunner(
+    task: Task,
+    objectFactory: ObjectFactory,
+): KotlinJvmCompilerRunner {
+    return objectFactory.newInstance<KotlinJvmCompilerRunnerImpl>(task)
+}
+
+fun createKotlinJsCompilerRunner(
+    task: Task,
+    objectFactory: ObjectFactory,
```

Comment:
Just passing the task should be enough from the API point of view.

Reply:
Will address in upstream patches. I'll probably also move those functions into proper KGP classes or factories.